### PR TITLE
Save and list user's presets for fields to be exported

### PR DESCRIPTION
**Pull request by funcionante**
For now, this targets a part of issue #221. It defines endpoints to allow using templates with predefined fields to export:
* Get the list of a user's presets: `GET /presets`
* Save a user's preset: `POST /presets?name=<name>&content=<content>`

### CODE REVIEW

File: dicoogle/src/main/java/pt/ua/dicoogle/server/users/PresetsHandler.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        if (!presetsDir.isDirectory()) {
+            return presets;
+        }
+
+        for (File presetFile: presetsDir.listFiles()) {
+            try {
+                // read all lines
+                String presetContent = new Scanner(presetFile).useDelimiter("\\Z").next();
+                presets.put(presetFile.getName(), presetContent);
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/users/PresetsHandler.java]: Won't this include the `.txt` extension in the keys?

File: dicoogle/src/main/java/pt/ua/dicoogle/server/web/servlets/search/PresetsServlet.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        String token = req.getHeader("Authorization");
+        User user = Authentication.getInstance().getUsername(token);
+        if (user == null) {
+            resp.sendError(401);
+        }
+
+        String presetName = req.getParameter("name");
+        String presetContent = req.getParameter("content");
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/web/servlets/search/PresetsServlet.java]: I forgot one more thing: could you try making an endpoint `/presets/<<username>>/<<preset-name>>` instead? The content should be sent at the request body. Users would only have access to their own presets, but administrators could have full access eventually.

**Comment by funcionante** [comment]
Thanks for the feedback. As suggested, now the endpoints are:

Get the list of a user's presets: `GET /presets/<username>`
Save a user's preset: `POST /presets/<username>/<preset-name>`

Some methods already used in other servlets, like `sanitizedSubpathParts()` and `sendError()`, may be useful globally. Should I extract and put them in another place?

**Comment by Enet4** [comment]
> Should I extract and put them in another place?

Yep, that's a good idea!

File: dicoogle/src/main/java/pt/ua/dicoogle/server/users/UserExportPresets.java [hunk 3]
```diff
@@ -49,38 +56,34 @@ public static boolean savePreset(String username, String presetName, String pres
...
-            Writer writer = new PrintWriter(presetFile);
-            writer.write(presetContent);
-            writer.close();
-
-        } catch (Exception e) {
-            presetFile.delete();
-            return false;
+        Writer writer = new PrintWriter(presetFile);
+        for (String field: fields) {
+            writer.write(field + "\n");
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/users/UserExportPresets.java]: Minor nit:

```
writer.write(field);
writer.write('\n');
```

File: dicoogle/src/main/java/pt/ua/dicoogle/server/users/UserExportPresets.java [hunk 4]
```diff
@@ -49,38 +56,34 @@ public static boolean savePreset(String username, String presetName, String pres
...
 
-        try {
-            Writer writer = new PrintWriter(presetFile);
-            writer.write(presetContent);
-            writer.close();
-
-        } catch (Exception e) {
-            presetFile.delete();
-            return false;
+        Writer writer = new PrintWriter(presetFile);
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/users/UserExportPresets.java]: Prefer try-with syntax here:

```
try (Writer writer = new PrintWriter(presetFile)) {
    // ...
}
```

File: dicoogle/src/main/java/pt/ua/dicoogle/server/users/UserExportPresets.java [hunk 5]
```diff
@@ -49,38 +56,34 @@ public static boolean savePreset(String username, String presetName, String pres
             return false;
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/users/UserExportPresets.java]: I think we should allow the user to update existing presets. Let's discard this check and let the file be overwritten.

This also means we don't need the return type to be `boolean`.

File: dicoogle/src/main/java/pt/ua/dicoogle/server/web/servlets/search/PresetsServlet.java [hunk 6]
```diff
@@ -65,24 +86,65 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+
+        // the format must be /<username>
+        if (pathParts.size() != 1) {
+            sendError(resp, 400, "Illegal plugin request URI: wrong resource path size");
+            return;
+        }
+
+        String username = pathParts.get(0);
+
+        // if the specified user is not the requester, he must have admin privileges
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/web/servlets/search/PresetsServlet.java]: s/he/they/ (_", they must have admin privileges"_). Not important, but it's best to keep users gender-neutral.

File: dicoogle/src/main/java/pt/ua/dicoogle/server/web/servlets/search/PresetsServlet.java [hunk 7]
```diff
@@ -65,24 +86,65 @@ protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws Se
...
+        if (!username.equals(user.getUsername()) && !user.isAdmin()) {
+            sendError(resp, 401, "Unauthorized access to this user.");
```
Comment (Enet4) [inline dicoogle/src/main/java/pt/ua/dicoogle/server/web/servlets/search/PresetsServlet.java]: The error code 403 might be more accurate here: the user is authenticated at this point, but does not have permission to access the presets.

**Comment by funcionante** [comment]
Now targets the whole issue #221. The changes to the webapp now allow to:
- load previously exported presets;
- save new presets (it is possible to override previous ones);
- choose from the list of existing fields, only supporting valid values;
- only export files with at least one field;
- only save presets with non-empty names;
- display a success message when a preset is saved.

It looks like this:

image

**Comment by Enet4** [comment]
Whelp, we have a performance issue here. The known list of tags is pretty big, and the default select component invokes a `filter` over the current list of candidates on each text input event. In my bench over here, some of these can take 200ms. By the looks of it, we could really use an index at the tag list storage or something similar.

screenshot_2017-12-18_17-13-06

I also think that the user should be able to specify custom tags even if they are not listed, because they might be private and not yet registered in the platform by the user (AFAICT the webapp doesn't facilitate this either!). But I'd like to hear more opinions (CC @tmgodinho @bastiao )

**Comment by funcionante** [comment]
Now the filtering is being done asynchronously. Also, as nobody opposed it, it is allowed to add custom tags.

**Comment by funcionante** [comment]
Thanks for the feedback. Now the unselection in the preset field should be correctly handled.

**Comment by bastiao** [comment]
I will handle that.

**Comment by bastiao** [comment]
Done.

**Pull request merged**